docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,519 @@
+if (args_ptr->table_id == OE_UINT64_MAX)
+{
+    ecall_table.ecalls = __oe_ecalls_table;
+    ecall_table.num_ecalls = __oe_ecalls_table_size;
+}
+else
+{
+    if (args_ptr->table_id >= OE_MAX_ECALL_TABLES)
+        OE_RAISE(OE_NOT_FOUND);
+
+    ecall_table.ecalls = _ecall_tables[args_ptr->table_id].ecalls;
+    ecall_table.num_ecalls = _ecall_tables[args_ptr->table_id].num_ecalls;
+
+    if (!ecall_table.ecalls)
+        OE_RAISE(OE_NOT_FOUND);
+}
+```
+
+**The Limitation of Per-EDL Tables**
+
+The workaround introduces per-EDL tables that address the limitation that
+an enclave cannot include multiple EDL files. However, this approach
+does not scale well when extending it to address the other limitation:
+multiple enclaves on the same host cannot import/use the same EDL file
+(e.g., including the same set of ECALLs).
+
+Considering an example with two enclaves, say `enclave_foo` and `enclave_bar`.
+Each enclave defines its own EDL file with a different number of ECALLs.
+In addition, both the EDL files import the same EDL file (the same `sample.edl` as above).
+As a result, the corresponding ECALL tables (defined in each `_t.c` file) include
+the wrapper function of `sample_ecall` but with different function id (because
+the number of ECALLs is different).
+Assuming that we use *weak symbol* to avoid the naming conflict of
+the wrapper function (i.e., only one implementation of the functions is picked),
+the invocation of `sample_ecall` to both enclaves ends up using the same
+table id and function id and potentially causing a mismatching;
+i.e., the enclave may use the wrong table id or function id to look up the ECALL.
+
+Proposed revision
+-----------
+
+To address two limitations more fundamentally, this document proposes to
+use a single table that includes ECALLs or OCALLs in multiple EDL files.
+This section describes the detail of the design and considerations.
+
+**Maintain Single Table for ECALL and OCALL**
+
+Instead of adopting the per-EDL tables, maintaining a single table (i.e., one
+OCALL table and one ECALL table) provides finer granularity to keep track
+of ECALLs or OCALLs from multiple EDL files that are either imported or directly included.
+To avoid duplicated entries, each ECALL or OCALL is registered by its hash of the name.
+  
+**Add the registration functions of ECALLs and OCALLs in `_t.c` and `_u.c`**
+
+Instead of using wrappers for `_t.c` and `_u.c` files, each `_t.c` or `_u.c` now
+includes functions that register the list of OCALLs/ECALLs by passing the function
+pointers and the hashes of function names. See the following example.
+  
+```
+static oe_ecall_func_t __oe_ecalls_table[] = {
+    (oe_ecall_func_t) ecall_oe_verify_report_ecall,
+    (oe_ecall_func_t) ecall_oe_get_public_key_ecall,
+    (oe_ecall_func_t) ecall_oe_get_public_key_by_policy_ecall,
+    (oe_ecall_func_t) ecall_oe_log_init_ecall
+};
+
+static uint32_t __oe_ecalls_hash_list[] = {
+    3802944385 /* CRC32 of oe_verify_report_ecall. */,
```

Comment:
Why use CRC32 here? Is it just calculated over the function name?

Your odds of hitting a natural collision in a 32-bit space over just ~10k unique values is non-trivial, so if that is the intended mechanism for uniqueness in a consolidated single table, then you should probably use a 64-bit value. GUIDs have long been used for non-string identification purposes (see OP-TEE's TA identifiers or COM GUIDs) for global uniqueness.

Reply:
How about CRC64 or MD5 (truncated)? The hash needs to be deterministic so that oeedger8r can generate the same hash for a function in both `_t.c` and `_u.c` every time. I think GUIDs is randomly generated.

Reply:
The proposal uses the name as an identifier now.